lib/AST/TypeJoinMeet.cpp

```diff
@@ -352,16 +345,130 @@ CanType TypeJoin::visitGenericFunctionType(CanType second) {
   return Unimplemented;
 }
 
+// Use the distributive law to compute the join of the protocol
+// compositions.
+//
+//   (A ^ B) v (C ^ D)
+// = (A v C) ^ (A v D) ^ (B v C) ^ (B v D)
+CanType TypeJoin::computeProtocolCompositionJoin(ArrayRef<Type> firstMembers,
+                                                 ArrayRef<Type> secondMembers) {
+  SmallVector<Type, 8> result;
+  for (auto first : firstMembers) {
+    for (auto second : secondMembers) {
+      auto joined = Type::join(first, second);
+      if (!joined)
+        return Unimplemented;
+
+      if ((*joined)->isAny())
+        continue;
+
+      result.push_back(*joined);
+    }
+  }
+
+  if (result.empty())
+    return TheAnyType;
+
+  auto &ctx = result[0]->getASTContext();
+  return ProtocolCompositionType::get(ctx, result, false)->getCanonicalType();
+}
+
 CanType TypeJoin::visitProtocolCompositionType(CanType second) {
+  // The join of Any and a no-escape function doesn't exist; it isn't
+  // Any. If it were Any, it would mean we would allow these functions
+  // to escape through Any.
   if (second->isAny()) {
     auto *fnTy = First->getAs<AnyFunctionType>();
     if (fnTy && fnTy->getExtInfo().isNoEscape())
       return Nonexistent;
 
-    return second;
+    return TheAnyType;
   }
 
-  return Unimplemented;
+  assert(First != second);
+
+  // FIXME: Handle other types here.
+  if (First->getKind() != TypeKind::Protocol &&
```

Comment:
You should be using `isa<>` instead of `getKind()`, I think

Reply:
Ah yes, I'll fix that up.

Reply:
Is that because isa<> strips sugar? I still can’t say when it would make sense to use getKind vs. isa<> with confidence... is getKind supposed to be used only by switch statements?

Reply:
In general that's a good reason to use `is<>` and `getAs<>`, but in this case we're working with `CanType`s so that shouldn't make a difference (`*`). It's generally shorter and more idiomatic to use these templated functions.